### ISSUE TITLE
Suggestion: "safe navigation operator", i.e. x?.y

### ISSUE THREAD

**Issue by RyanCavanaugh** [microsoft/TypeScript#16]
# Current Status

 * The TC39 proposal is now at stage 3 (🎉🎉🎉🎉🎉)
 * Implementation is in progress
 * You can expect this feature in TypeScript 3.7
 * We'll update here when it's available in a nightly build
 * Holding off on Optional Call until its semantics are clarified at committee

Open questions

 * What special-casing, if any, should `document.all` get?

--------

C# and other languages have syntax sugar for accessing property chains where `null` (or in our case, `undefined`) might be encountered at any point in the object hierarchy.

``` TypeScript
var x = { y: { z: null, q: undefined } };
console.log(x?.y?.z?.foo); // Should print 'null'
console.log(x?.baz); // Still an error
console.log(x.y.q?.bar); // Should print 'undefined'
```

Need proposal on what exactly we should codegen, keeping in mind side effects of accessors.

_____
***Edit by @DanielRosenwasser February 27, 2018:** This proposal is also called the "null propagation" operator.*

**Comment by basarat** [comment]
The [codeplex issue had quite a number of votes](https://typescript.codeplex.com/workitem/118) (61)

I really badly need this to ease the pain of using `atom` for [atom-typescript](https://atom.io/packages/atom-typescript).

It is very idiomatic in coffescript code (although I would like it not to be as popular as determinism is better than a fudgy `?`). Open any coffescript file, especially one that works with the DOM directly like space-pen (where functions can run after the view is destroyed or before the view is attached) and you will find a gazillion `?` usages. e.g. this file has `16` https://github.com/atom-community/autocomplete-plus/blob/f17659ad4fecbd69855dfaf00c11856572ad26e7/lib/suggestion-list-element.coffee

Again I don't like that I need this, but its the state of JavaScript, and I'd rather `?` than a million `if( && fest ) { then }`

But I really really need it to keep my code readable. Its also very common to *need* this when you are waiting for an XHR to complete and angular runs its digest loop.

**Comment by basarat** [comment]
Okay now I have read the thread and see why we are waiting. I understand *sigh*.

**Comment by basarat** [comment]
`we'd have to somehow make x, y, z, and foo each evaluate at most once.`

coffeescript does do some optimizations e.g. [stores intermediate access results](http://coffeescript.org/#try:x%3F.y%3F.z%3F.foo):

```js
var _ref, _ref1;

if (typeof x !== "undefined" && x !== null) {
  if ((_ref = x.y) != null) {
    if ((_ref1 = _ref.z) != null) {
      _ref1.foo;
    }
  }
}
```

(I strongly feel the `undefined` check is unnecessary for typescript : as we should have a `var` init typechecked by typescript)

**Comment by rayshan** [comment]
+1

**Comment by basarat** [comment]
In news today, Dart is getting official support for it : https://github.com/gbracha/nullAwareOperators/blob/master/proposal.md

**Comment by Back-io** [comment]
Very important feature.

Possibly an off-the-wall idea but the codegen for this feature could be done very easily without side effects if everyone decided that it would be OK to handle the feature with keyed property access:
```
if (aaa?.bbb?.ccc) {}
```
Could compile to
```
if (__chain(aaa, "bbb", "ccc")) {}
```
A `__chain` function would have to be emitted similar to `__extends`. The `__chain` function could just iterate through the `arguments` array, returning `null` when the upcoming member is not `instanceof Object` or does not contain the member name. Function calls could be handled by passing in an array as a parameter, so ...
```
if (aaa?.bbb?.ccc?(1, 2, 3)) {}
```
Could compile to
```
if (__chain(aaa, "bbb", "ccc", [1, 2, 3])) {}
```

This would also keep the generated JS reasonably idiomatic, even for long chains.

**Comment by kevinb7** [comment]
If ```aaa?.bbb?.ccc?``` returns the value of ```a.b.c``` if all props exist and it happens to be a function, then couldn't
```
if (aaa?.bbb?.ccc?(1, 2, 3)) {}
```
compile to
```
if (__chain(aaa, "bbb", "ccc")(1, 2, 3)) {}
```
?

**Comment by Back-io** [comment]
@metaweta: Your solution is only checking for void 0 ... doesn't that kind of defeat the purpose of the feature?

What about this:

```
var result = one?.two?.three;
```
Generates:
```
var $a, $b, $c;
var result = $a = one, $b = $a ? $a.two : null, $b ? $b.three : null;
```

I'm pretty sure this handles all the cases.

(Minor downside here of unexpected local variables being emitted ... could be wrapped in an IIFE maybe)

**Comment by Back-io** [comment]
@kevinb7: What happens if `ccc` isn't a function? With the code you've described, `__chain` would always have to return a valid function, or a TypeError would be emitted.

**Comment by metaweta** [comment]
@Back-io Replace === with == to include null.  Your solution fails for looking up properties of falsey values like empty string and zero.

**Comment by Back-io** [comment]
@metaweta: No it doesn't: http://jsfiddle.net/25LppbL6/

**Comment by Back-io** [comment]
Also, I would imagine the TS team isn't crazy about using loose equality due to the fact that some people's linters warn against it.

**Comment by metaweta** [comment]
@Back-io Yes it does: http://jsfiddle.net/25LppbL6/2/

**Comment by metaweta** [comment]
@Back-io Regarding null, I'm not sure what the intended semantics of a.?b is.  If it's "If the property b is defined then use it", then my code is almost correct.  The only way you'd get null is if it's assigned to be null, because lookups of nonexistent properties return undefined.  It does fail to catch the case where the property exists but is set to undefined.  To be completely correct, it would check with Object.hasOwnProperty() instead of comparing to void 0 === undefined.

If the semantics are "if the property b is truthy then use it", your code is fine and to some extent matches JS idiom.

**Comment by Back-io** [comment]
Umm ... unless I'm missing something ... the changes you made to the fiddle just further prove me right ... `var result` is still `undefined` in all 3 cases. And I'm not sure what case you're trying to bring forward by extending the primitive prototypes ...

**Comment by Back-io** [comment]
I'm fairly certain that the behavior of the feature would be to short circuit with `void 0` instead of generating an error. (I like your idea above that it should always return void 0 in the case of a missing member).

**Comment by metaweta** [comment]
What is the intended output of `''.?toString`?

**Comment by kevinb7** [comment]
> @kevinb7: What happens if ccc isn't a function? With the code you've described, __chain would always have to return a valid function, or a TypeError would be emitted.

Good point.

**Comment by Back-io** [comment]
@metaweta: I would imagine most people would expect a reference to the `toString` function, so I see where your point, It does break down somewhat in the case where you're accessing the prototype members of 0, false, or ''.

**Comment by gdi2290** [comment]
:+1: 
Angular 2 added Elvis Operator to their template syntax

**Comment by basarat** [comment]
> What is the intended output of ''?.toString?

If you meant `''?.toString()` it would be : 

```ts
if ('' != null) {
  ''.toString();
}
```

[Sample](http://coffeescript.org/#try:''%3F.toString())

**Comment by basarat** [comment]
> Maybe we could put together a strawman proposal for this and then have a reference implementation

@kevinb7 it already exists : http://wiki.ecmascript.org/doku.php?id=strawman:existential_operator

**Comment by basarat** [comment]
I did some quick tests to implement this on top of `PropertyAccessExpression` as its special case but didn't work out well as we really need `?.` to be right associative (instead of left associative like `.`) otherwise the emitter becomes needlessly complex. 

There is comment by BrendenEich [here](https://github.com/JSFixed/JSFixed/issues/71#issuecomment-9325586) that reflects this as well.

**Comment by zlumer** [comment]
> I did some quick tests to implement this on top of PropertyAccessExpression as its special case but didn't work out well as we really need ?. to be right associative (instead of left associative like .) otherwise the emitter becomes needlessly complex.

@basarat can you please elaborate on that? An example of the difference between right associative and left associative `?.` would be very helpful to me.

**Comment by basarat** [comment]
@zlumer 

> An example of the difference between right associative and left associative ?. would be very helpful to me.

`.` is left associative so the the expression `foo?.bar?.baz` becomes in AST (if we treat `?.` the same):

```
                    // foo.bar.baz = PropertyAccessExpression
                    //   .expr foo.bar =  PropertyAccessExpression
                    //     .expr foo = Identifier
                    //     .name bar = Identifier
                    //   .name baz = Identifier
```

The JavaScript emit needed is 

```
foo != null ? (ref_1 = foo.bar) != null ? ref_1.baz() : void 0 : void 0;
```
Its just easier to do this emit (especially *recursively*) if we had the following in the AST: 
```
                    // foo.bar.baz = PropertySafeAccessExpression
                    //   .name foo =  Identifier
                    //   .expr bar.baz = PropertySafeAccessExpression
                    //      .expr bar = Identifier
                    //      .name baz = Identifier
```

Just think of *how would you convert the first AST to JavaScript* and the complexities would be clearer. Hope this helps :rose:

**Comment by DanielRosenwasser** [comment]
@zlumer, to add to what @basarat said, I'll just put out the example `1 + 2 + 3`.

When parsing, we have to decide which of these operations will happen first.

If `+` is left-associative, this will be interpreted as `((1 + 2) + 3)`.

If `+` is right-associative, this will be interpreted as `(1 + (2 + 3))`.

You might question whether this would actually make a difference. In JavaScript it would! Consider the example `"hello" + 2 + 3`.

* Left-associative: `(("hello" + 2) + 3) => ("hello2" + 3) => "hello23"`
* Right-associative: `("hello" + (2 + 3)) => ("hello" + 5) => "hello5"`

(For the record, JavaScript/TypeScript use left-associativity for the `+` operator.)

**Comment by DanielRosenwasser** [comment]
@basarat, my understanding of what @BrendanEich said (and he can correct me if I'm wrong - sorry for the ping!) is *not* that `?.` is right-associative, but that CoffeeScript special cases property accesses on the right of the `?.` to be right-associative. For example, it will parse

```CoffeeScript
o.p?.q.r.s
```

as

```CoffeeScript
((o . p) ?. (q.(r.s))) # or something close to this
```

instead of

```CoffeeScript
((((o . p) ?. q) . r) . s)
```

because it's easier to emit.

Our AST needs to lend well to sane semantic analysis, so our emitting procedure can afford to be slightly more complex to suit this need.

**Comment by zlumer** [comment]
@basarat @DanielRosenwasser thank you for the explanations. So far I understand it, but I'm still not certain about one thing.
The left-associative `?.` is pretty obvious and expected:
```js
foo?.bar?.baz
```
Becomes
```js
var ref = ((ref = foo) == null) ? null : ((ref = ref.bar) == null) ? null : ref.baz;
```
But I don't understand at all how would a right-associative `?.` work. Can you please provide an example?

**Comment by basarat** [comment]
> But I don't understand at all how would a right-associative ?. work. Can you please provide an example

@zlumer The runtime behaviour will be left associative. I was just talking about the AST as DanielRosenwasser also clarified: `is not that ?. is right-associative, but that CoffeeScript special cases property accesses on the right of the ?. to be right-associative`.

> Our AST needs to lend well to sane semantic analysis, so our emitting procedure can afford to be slightly more complex to suit this need.

@DanielRosenwasser thanks for the feedback :rose:

**Comment by zlumer** [comment]
@basarat thank you, suddenly it all became clear :smiley:

**Comment by schungx** [comment]
Similar to @basarat's Feb comment, I.... *sigh*...

However, if you think about it, 99% of the use cases of this will be to check against a null pointer to an object.  Frankly speaking, who does `x?.b?.c` when `x` is a `number`?  There just aren't many real-life use cases for *long* chains when we are *not* talking about an object (with the possible exception of `string`).  For short chains, I think we can live with `x && x.b` or `x === 0 ? null : x.b`.

So, can we, sorta, say `?.` only works on object types?  Any other type throws a syntax error.  And disallow function calls in side the chain.

Then the whole thing transcribes to `a && a.b && a.b.c`.

**Comment by Back-io** [comment]
@schungx What if a member in the chain is an "any" type? Disallow completely? Or just let it through and hope for the best?

**Comment by schungx** [comment]
Well, my suggestion?  Disallow completely.  Inelegant as hell, I know...  :-)

But my rationale:

1. This is a short-hand, so if somebody is using `any`, just use long-hand.
2. If somebody is using TypeScript, most likely you use it for the typing support, so hopefully he/she won't have many `any`'s around!
3. `any` really should be handled with care.  Allowing usage of such short-hands with a flexible type like `any` is really asking for bugs to happen.  In my opinion, `any` should be as limited as possible.  It is sort of like C's `(void *)` -- the fact that you're handed a nuke doesn't mean you must trigger it just because you can!

**Comment by dotnetwise** [comment]
This would be amazing operator!!

```js
var error = a.b.c.d; //this would fail with error if a, b or c are null or undefined.
var current = a && a.b && a.b.c && a.b.c.d; // the current messy way to handle this
var typeScript = a?.b?.c?.d; // The typescript way of handling the above mess with no errors

```

However I propose a more clear one - as not to confuse ? from the a ? b : c statements with a?.b statements:

```js
var x = a..b..c..d; //this would be ideal to understand that you assume that if any of a, b, c is null or undefined the result will be null or undefined.
```

Two dots, means if its null or undefined stop processing further and assume the result of expression is null or undefined. (as d would be null or undefined).

Two dots make it more clear, more visible and more space-wise so you understand what's going on.

What do you think folks?

**Comment by hsablonniere** [comment]
Very interesting. :+1:

**Comment by schungx** [comment]
IMHO, two dots will be more confusing.  There are languages where two dots represent a range (e.g. `1..4`) and TypeScript may add this feature in the future.

A question mark also has the semantic meaning of uncertainty or conditional and two dots do not convey the same meaning.

**Comment by hsablonniere** [comment]
@schungx Fair enough, but it would help for the weird possibilities like this : `a?['b']` or this `a?()`.

**Comment by AshCoolman** [comment]
+1 `?`

`a?['b']` and `a?()` behave nicely in coffeescript, and they look good to me.

But then again, I might just be coffeescript-blind.

**Comment by p-mcgowan** [comment]
+! Ruby just implemented the existential operator https://twitter.com/mikepack_/status/657229703443451904. Typescript needs this too, regardless of the specific syntax.

**Comment by andriniaina** [comment]
This would be make code cleaner 👍

**Comment by aguzubiaga** [comment]
+1 Need this!

**Comment by svekl** [comment]
Please, implenent ?. operator, as C# does

**Comment by pavlovt** [comment]
+1 I've been missing this for a long time

**Comment by osi-oswald** [comment]
+1
so ugly to write: someVariable && someVariable.someMember
when you could write: someVariable?.someMember

**Comment by maku** [comment]
+1, this would be great.... (most wanted feature for me)

**Comment by aere-github** [comment]
+1, it's a good idea! 
Only, if the object is complicated, the expression will be full of ?. after each property (var result=myValue?.a?.b?.c?.d?.e;) when I need to recieve a value of the last one (var result=?myValue.a.b.c.d.e;).

**Comment by bryanmig** [comment]
+1 - This is arguably one of the greatest features of CoffeeScript and is by far my team's most desired TypeScript feature after converting much of our code from CS to TS.

**Comment by falsandtru** [comment]
+1 however, this is too complex:

```ts
var x = { y: { z: null, q: undefined } };
var z: x|y|z = x?.y?.z;
```

I like this:

```ts
var x = { y: { z: null, q: undefined } };
var z: z|void = x?.y?.z;
```

**Comment by schungx** [comment]
`x?.y?.z`'s type is always the type of the `z` field.

**Comment by gaurav21r** [comment]
+1 This would go well with Typescript's vision of easing development of large scale complex JS Projects.

Any updates on this? Is this a case of the community voting for this feature for it to be considered? Or has it been considered but there are some engineering challenges?

**Comment by DanielRosenwasser** [comment]
No updates so far because introducing new expression-level syntax is dangerous without some sort of proposal from the ECMAScript committee.

See https://github.com/Microsoft/TypeScript/issues/16#issuecomment-57645069.

**Comment by kitsonk** [comment]
Here is a more current discussion on existential operators (lots of thoughts, but doesn't look like a lot of action):

* [Existential Operator/Null Propagation Operator](https://esdiscuss.org/topic/existential-operator-null-propagation-operator)

**Comment by msklvsk** [comment]
where should I write "+1" to help bring it to ES?

**Comment by kitsonk** [comment]
@msklvsk [ESDiscuss](https://esdiscuss.org/)

**Comment by RyanCavanaugh** [comment]
Closing this for now. Since there's not really anything TypeScript-specific that would require this at expression level, this kind of big operator change should happen at the ES spec committee rather than here.

The general tripwires for re-evaluating this would be a concrete ES proposal reaching the next stage, or a general consensus from the ES committee that this feature wouldn't happen for a long time (so that we could define our own semantics and be reasonably sure that they would "win").

**Comment by luisrudge** [comment]
life sucks

**Comment by loretoparisi** [comment]
👍

**Comment by nisimjoseph** [comment]
+1

**Comment by RyanCavanaugh** [comment]
Deleting standalone :+1:s. Please use the GitHub reactions feature or send flowers and candy to your nearest TC39 representative.

**Comment by algesten** [comment]
Once I got used to it in coffeescript and Swift, there's no going back. TS is dead to me as it currently stands.

**Comment by RyanCavanaugh** [comment]
:ghost:

**Comment by loretoparisi** [comment]
@algesten I can agree about `swift` if we consider the language itself, not sure about long term `coffescript` support (I use CoffeScript in production projects). We may agree to be sure of language analytics trends as for June 2016 like [this](http://redmonk.com/sogrady/category/programming-languages/) where we can clearly see what's going on to `coffescript` recently, while `TypeScript` has seen the fastest growth in the past years:


> TypeScript: Outside of Go or Swift, the fastest growing language we’ve observed in recent years is TypeScript. The Microsoft-backed JavaScript superset and Angular 2 foundation has made significant gains for the second consecutive quarter, jumping from #31 to #26. That was the biggest single change in any Top 30 language, and the second largest jump overall (Standard ML, 7 spots). At #26, in fact, TypeScript is now tied with Erlang, one spot behind Powershell and four behind CoffeeScript, which is just outside the Top 20. The question facing the language isn’t whether it can grow, but whether it has the momentum to crack the Top 20 in the next two to three quarters, leapfrogging the likes of CoffeeScript and Lua in the process.

**Comment by yuvalrakavy** [comment]
With the new typescript 2.0 strict null (undefined) checking, it is a must, because otherwise you cannot use function chaining!

**Comment by aluanhaddad** [comment]
This should be in the next version of ECMAScript. Considering this would be even more useful in vanilla JavaScript than TypeScript and considering that the implementation would naturally be a check for undefined or null as opposed to a truthy check, it should be trivial for TC39 to add. 

The simple implementation and one which would be idiomatic in JavaScript, would simply be to short circuit when encountering either null or undefined, and return undefined. 

@bterlson does this make sense? What are the odds of such a proposal being accepted?

**Comment by yuvalrakavy** [comment]
One of the great benefits of typescript is "the future is given to you today". This is one of the features that I was surprised that are not already there.

 I hope that it will be added soon as function chaining is popular idiom and will strict null checking it is no longer work unless the ?. operator is added.

TS 2 wonderful null checking made it from a nice to have to must have!

**Comment by cervengoc** [comment]
Reading through this thread I'm kind of surprised how this doesn't get more attention from the maintainers. 

In an ideal world TypeScript should lead the way for ES and not reversely. Seriously, where would be TypeScript now if the TS team would always have waited for ESx to propose and finalize a feature or a syntax?

This syntax is really a "must have", as others have pointed it out. It even received some good proposals in this thread so far. I think the implementation should match these expectations.

- In general an expression `a?.b` should be valid at compile-time if and only if `a.b` is valid.
- It should evaluate each expression in the chain only once.
- It should be short-circuit.
- If execution reaches a mid-expression with `null` or `undefined` value then that value should be the return value.
- If execution reaches a mid-expression with a value which is invalid for further chaining (like `number`s) and it didn't produce compile errors, then it should fail at runtime.
- A empty string should be always treated as a valid expression, as it can be chained.

What do you think the parts are which can lead to disagreements when ES will specify it?

**Comment by osi-oswald** [comment]
For `a?.b` I do not see any conflicts with existing (and future?) syntax. If the parser finds the token `?.` it can treat it as 'safe navigation operator' (with expectations as described by @cervengoc).

Syntax conflicts arise only when allowing `a?(b)` and `a?[b]`, since these could also be interpreted as the start of a ternary `?:` operator expression. But for starters I think these could be put aside and supporting just the `a?.b` syntax would already make a lot of developers happy!

**Comment by RyanCavanaugh** [comment]
> In an ideal world TypeScript should lead the way for ES and not reversely.

In terms of expression-level syntax, we absolutely disagree. There's a committee driving a standard for a reason -- not so that one-off actions by one player can unilaterally decide the future of JavaScript.

> This syntax is really a "must have", as others have pointed it out. 

If this is a must-have for TypeScript, it's a must-have for JavaScript as well! Again, **take your concerns to the ECMAScript committee**.

> What do you think the parts are which can lead to disagreements when ES will specify it?

At a minimum, I think there will be disagreement about whether the syntax short-circuits to `null` or `undefined` when encountering those values, or always short-circuits to `undefined`. There is also going to be some contention around whether or not some form of bracketed syntax is supported. There's also the question of what the behavior of `a?.b.c` is. There's also the question of `?.` vs `.?` vs `a.b?`. There's the question of what the effect of this on the `delete` operator is.

The [ES DIscuss thread on this](https://esdiscuss.org/topic/existential-operator-null-propagation-operator) is over 100 comments long. There's no lack of ambiguity! It's easy to look at one example in isolation and think there can't be tons of corner cases. There are. That's probably why no one has championed this at TC39 yet and *also* why we're not rushing to add a feature that has lots of ambiguity around how it should behave.

**Comment by cervengoc** [comment]
I apologize, I didn't read through the mentioned thread, I will definitely have a look at it to see more.

We see this a bit differently. About the committee, in my honest opinion this is one of the biggest reasons why JavaScript will never be *good*. Just for an example, many of the most successful software what I've seen (like Total Commander or IrfanView) are successful because they are maintained and designed by ONE person, and not by a *committee". Of course this is not a completely correct example. But I'm almost sure, that if for example you alone would have designed the complete ES6, then the world would be a better place now.

Also, the ambiguities which you've mentioned are in 99% *theoretical* things and kind of irrelevant from the developer's side. Who would care about what it returns, `null` or `undefined`? Just pick one, and we will use it like that. 

All in all, you and that committee are on a different side than most of us, and things from that side are usually more complex than they really are. And this can lead to some contra-productivity to say the least. Don't take it personally, but according to my experience in general, some people would be better come out of the conference room more often and have a look at some code.

Of course no offense, don't take anything personally, I have a huge respect to you and all the TS team, because you did revolutionize many developers' client-side development including myself, and thank you for all your work. We're just a bit disappointed about this specific one I guess.

**Comment by cervengoc** [comment]
One last thought. I've ran through the mentioned ES thread, and one thing is completely sure: they are overcomplicating it. They want to design something which is good for every scenario.

I personally would be completely satisfied and happy with a conditional **member access* operator. Don't need conditional invocation, conditional index signature, don't need to support every fancy scenario which is valid JS code. That's all. But instead, they will likely keep sitting there and discuss how to do everything at once, which is a great plan, but we will have nothing at the end of the day.

**Comment by kitsonk** [comment]
> All in all, you and that committee are on a different side than most of us, and things from that side are usually more complex than they really are.

I don't think you are reflecting the true status of Ryan or TC39 accurately.  Ryan, and the TypeScript team, have set out very clear [design goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals) for TypeScript.  One of the original and still very current goals is that TypeScript is a superset of JavaScript.  Not the language people would like it to be (e.g. Dart, Haxe).  When it come to syntax, the TypeScript team have learned the hard way the cost of pre-inventing it (e.g. modules).  We are headed headlong into a challenge with private members of classes too, where the proposed ES syntax is totally incompatible with the syntax TypeScript uses.  Why? Because what may look uncomplicated on the surface is impossible to achieve given the run-time challenges of a language.

TC39 has "saved" JavaScript in my opinion.  ES4 was abandoned, not because it lacked good, innovative ideas, but because it would break the internet.  TC39 got themselves into shape, they have shared and been totally open in their discussions and how they make decisions and delivered us ES2015, which is a lot like ES4, but didn't break the internet.  It is amazing that we essentially have JavaScript run-times that run code from 10 years ago just fine, but support many significant improvements to the language.  ES2016 was the lull before the storm.  ES2017 has a "reasonable" amount of functionality and change and a clear governing process that heads the right direction.

So being on the "different side" of things clearly has worked, in my opinion.  Which beats expediency of "must have" features.

**Comment by cervengoc** [comment]
@kitsonk I didn't mean "different side" negatively, and especially I didn't mean to degrade the work what was put into TypeScript or ES6. What's more, the best thing in TypeScript IMO is that it really had and has a clear design goal, and it's well protected against becoming a havoc like many other opensource stuff.

I just wanted to tell that this very feature is a clear example of where a group of genius people will end up overthinking and overcomplicating things, instead of just going the easy simple way, and accept some limitations like not supporting invocations or index signatures, etc. Someone in that forum even suggested using this syntax in assignments, which is kind of crazy. I still think that this phenomenon is contra-productive in this sense. 

I understand that on your side it's a pain that for example private members became incompatible with the final ES6 concept. But on the other hand, we HAD it. Way before ES6. And that's the main point from our side. Roughly speaking, we don't care about how you manage to emit the appropriate code for it, we're just happily using it. Same with modules, and everything. We (or at least I) didn't see those pains what you're talking about it, we were always happy with private members or modules.

This particular feature is in CoffeScript as I read about it. Why do we, simple developers always have to make compromises when choosing a platform/library/plugin, etc.? I mean **always**. This is kind of annoying. Here we have a great language, which has great potential, which had completely leaves every other participants behind (including ES!), and which has successfully revolutionized a huge part of client-side development, and when it comes to this "simple" feature (I mean the member access part at least), we hear that it won't be implemented until ES commits to this. This is kind of annoying.

**Comment by gisenberg** [comment]
I wanted to give folks a quick heads up that this feature moved from stage 0 to stage 1 at today's TC39 meeting.

Relevant commit: https://github.com/tc39/proposals/commit/cb447642290a55398d483f5b55fb7f973273c75d
Meeting agenda: https://github.com/tc39/agendas/blob/master/2017/01.md

**Comment by luisrudge** [comment]
wow! that's huge!

**Comment by RyanCavanaugh** [comment]
Also worth a link to https://github.com/claudepache/es-optional-chaining

**Comment by RyanCavanaugh** [comment]
Some "surprises" I see here (not saying I disagree, just things that we probably would have done differently if we had done this earlier):

* `null` is not produced from an `a?.b` expression: when `a` is `null` this produces `undefined` instead
* Propagation in chained dots: `a?.b.c.d` will not throw if `b` and `c` properties are `undefined`
* Propagation *in presence of parens*: even `(a?.b).c` will not throw if `b` is undefined
* Propagation even occurs on method calls: `a?.b.c().d` will return `undefined` if the `c` invocation returns `null`
* `delete` operator is supported
* Bracketing syntax `a?.[x]` is supported
* Function call syntax `func?.(...args)` is supported, even for non-method calls (!)

**Comment by gisenberg** [comment]
I'd expect to see those areas between now and stage 2.

**Comment by algesten** [comment]
I think coffeescript got it right.

a?.b.c throws if b is undefined.

a?() and a?[0] are both good.

**Comment by jaredru** [comment]
> * Propagation in chained dots: a?.b.c.d will not throw if b and c properties are undefined
> * Propagation in presence of parens: even (a?.b).c will not throw if b is undefined
> * Propagation even occurs on method calls: a?.b.c().d will return undefined if the c invocation returns null

Those points don't seem accurate to me.  From the proposal:

```js
a?.b.c().d      // undefined if a is null/undefined, a.b.c().d otherwise.
                // NB: If a is not null/undefined, and a.b is nevertheless undefined,
                //     short-circuiting does *not* apply
```

**Comment by RyanCavanaugh** [comment]
Wow, I totally misread that. You're right. Updating

**Comment by LestaD** [comment]
@algesten from original proposal:

```js
a?.()

b?.[0]
```

**Comment by algesten** [comment]
sweet. the operator can sort of be though of as `?.` then.

**Comment by gisenberg** [comment]
There's some additional conversation happening over here: https://github.com/estree/estree/issues/146

**Comment by karol-depka** [comment]
A quote that could apply well: «Make simple things **easy**, and difficult things **possible**». Thus perhaps, support the most common cases well, while skip (at least initially) the complicated/rare cases, while they are still allowed to be "done manually" with longer (existing) syntax.

**Comment by mlesk** [comment]
Approach taken by facebook [idx: The Existential Function](https://facebook.github.io/react-native/blog/2017/03/13/idx-the-existential-function.html)

**Comment by cedvdb** [comment]
Just my two cents 

```
let a = b?.c?.d?.e;
```
to:
```
let a;
try{
   a = b.c.d.e;
}catch(e){
   a = undefined;
}
```

**Comment by RyanCavanaugh** [comment]
@cedvdb totally different semantics -- exceptions thrown in getters should not cause coalescing

**Comment by cedvdb** [comment]
@RyanCavanaugh yeah.. I didn't think this through.

**Comment by MgSam** [comment]
Is this on the radar for implementation now, or will the TS team wait for the ES proposal to move further along?

**Comment by mhegazy** [comment]
It is on our short list; we still have a few questions/concerns but you should see movement on this in the comming months.

**Comment by RyanCavanaugh** [comment]
Not sure where @mhegazy 's comment came from - the number of open questions on the TC39 proposal is too large for us to do meaningful work here. Specifically questions around how `null` and `undefined` interact and which syntax is actually supported need to be addressed first. Stage 2 is an absolute bare minimum and we'd prefer stage 3 given the impact on runtime behavior.

**Comment by zh99998** [comment]
does this simply code works?
```JavaScript
a == undefined ? expression : undefined
```
`expression` means a.x, a[x], a(x) 

then `a?.b?.[c]?.(d)` will generate
```
a == undefined ? (a.b == undefined ? (a.b[c] == undefined ? a.b[c](d) : undefined) : undefined) : undefined
```

seems will pass through all RyanCavanaugh's rule

delete could also generate in expression place

**Comment by kitsonk** [comment]
@zh99998 you _have_ to hate `==` because `''` and `0` are also equated.  It almost is argued that the runtime behaviour should be some sort of check on `(typeof value === 'object' || typeof value === 'function' || typeof value === 'symbol') && value !== null`, which is now getting to be rather complex.

As @RyanCavanaugh said, it is unlikely to progress until the [TC39 Proposal](https://github.com/tc39/proposals) for it progresses to at least stage 2 or 3.

**Comment by zh99998** [comment]
I see `==` only equals `null` and `undefined` as
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness

and test passed in chrome console:
```
'' == undefined
false
0 == undefined
false
```

**Comment by cedvdb** [comment]
undefined only coerce to null and vice versa. No other value is equal to undefined or null.

You might be confused with falsy values. 0 and "" are indeed falsy, but never equal null nor undefined.

**Comment by jehugaleahsa** [comment]
I see a lot of attempts using ternary operators, but any implementation that refers to the same property multiple times could result in unexpected side effects.

Fortunately, JavaScript has IIFE's, so you can store the result of an accessor in a function parameter, refer to it as much as you want and never evaluate it more than once. In my example below, I create a `coalesce` lambda that can be called multiple times in place of an expression containing `?.` operators. 

One thing the language implementors would need to concern themselves with are array and function calls in the middle of an expression. I think the compiler could handle detecting those situations and simply append those expressions onto the end of the coalesce call.

```typescript
const coalesce = (x: any, y: string) => x == null ? null : x[y];

const x = {
    y: {
        z: "Hello, World!!!"
    },
    f: () => "Foo!",
    a: ["Array!"]
};

// x?.y?.z
const value1 = coalesce(coalesce(x, 'y'), 'z');

// x?.f()
const value2 = coalesce(x, 'f')()

// x?.a[0]
const value3 = coalesce(x, 'a')[0]
```

Note there's no need for an actual global named "coalesce". This code could be directly inlined into any expression. It might cut down on bloat giving it a name, though.

I am not too concerned with the syntax used for the operator or if the language implementors want to wait for a standard. Just thought I would show off another approach.

**Comment by jehugaleahsa** [comment]
This also brings up the need for a `??` operator. In C#, this replaces any `null` expression with whatever is to the right.

```csharp
string x = null ?? "Hello";
````

In JavaScript, it is more idiomatic to use `||` to replace "falsey" values with the value on the right. 

```javascript
var x = null || "Hello";
```

Unfortunately, truthiness captures too many edge cases (`0`, `false`, etc.). Working with a null coalesce (`?.`), operator, you'd want something specific to `null`-ness.

```typescript
const x = { y: "" };
const result1 = x?.y || "default";  // I'd expect "default"
const result2 = x?.y ?? "default";  // I'd expect "" 
```

**Comment by bschlenk** [comment]
@jehugaleahsa, in your example with coalesce there would have to be a way to prevent function calls and member access from happening if the check before returned null. In the x?.f() example, f shouldn't be called if x is null.

**Comment by jehugaleahsa** [comment]
@bschlenk I don't think I agree. I think it should fail with a message like `null is not a function`, but that's not up to me, I guess.

**Comment by noppa** [comment]
The recent surge in comments detailing possible ways to implement the operator is a bit weird.
The implementation is probably a solved problem.

If you are interested, there's [the idx library](https://github.com/facebookincubator/idx), which closely resembles the behavior of the `?.` operator, although it also ignores lots of details of the planned operator. Anyway, their [specs for the compile output](https://github.com/facebookincubator/idx/blob/master/packages/babel-plugin-idx/src/babel-plugin-idx.test.js) might be of interest for anyone wondering about the ways this stuff could be implemented.

TS could output something like that or it could output something completely different, but I don't think that's what we are waiting here. It has been said many times here that TS won't get the operator until the ES proposal moves to one direction or to the other.

 It's the semantics that still has some unknowns, listed [here](https://github.com/claudepache/es-optional-chaining/issues) and [here](https://github.com/tc39/proposal-optional-chaining/issues).

**Comment by RyanCavanaugh** [comment]
Rest assured we'll be implementing this correctly and don't need another 100 comments here to figure out what `||` and `? ... : ...` do. As @noppa noted we're just waiting for the ES spec to finalize.

**Comment by STRML** [comment]
https://github.com/babel/babel/pull/5813 (accompanying [babylon PR](https://github.com/babel/babylon/pull/545)) was just merged into Babel's `preset-stage-1`. The spec is still at stage 1, of course, but this will help move it forward.

**Comment by abenhamdine** [comment]
Maybe I'm wrong, but I didn't see any link to the tc39 proposal in this thread, so here it is : https://github.com/tc39/proposal-optional-chaining

**Comment by RyanCavanaugh** [comment]
FYI optional chaining will be at TC39 next week https://github.com/tc39/agendas/blob/master/2017/07.md

**Comment by gisenberg** [comment]
@jehugaleahsa I think you're right and I'll be presenting `??` at TC39 next week. You can follow the proposal here: https://github.com/gisenberg/proposal-nullary-coalescing

**Comment by markwhitfeld** [comment]
I see that optional chaining was covered at the TC39... What was the verdict?
Hopefully it was enough to move this forward in Typescript 😉

**Comment by alangpierce** [comment]
@markwhitfeld From the [notes summary](https://github.com/rwaldron/tc39-notes/blob/master/es8/2017-07/summary.md):
Optional Chaining Operators: remains on Stage 1, will come back later with clearer definitions for various options and answers to feedback

Full notes here: https://github.com/rwaldron/tc39-notes/blob/master/es8/2017-07/jul-27.md#13iia-optional-chaining-operator

There are still open questions around how the operator should behave, so seems like it's not yet in a state where it can be added to TypeScript.

**Comment by Tiedye** [comment]
From those notes, it looks like the committee has no idea how the options presented work, I would've thought they'd have taken to learn about the proposals they were planning on discussing, before discussing them.  I guess its going to be a while longer until I can enable strict null checks.

I very much hope they go with option 2, it makes the most sense and is just as capable as option 1 when used with parenthesis.

**Comment by notsonotso** [comment]
I want this now, or I'll put my sad face on.

**Comment by frankfvb** [comment]
15 Jul 2014 - 4 Sept 2017, nothing yet

**Comment by kitsonk** [comment]
@frankfvb you have clearly not read the issue.

There has been a lot of discussion which has lead the core team to believe that it would be imprudent to implement at this juncture until there is further progress on the [ECMAScript proposal](https://github.com/tc39/proposal-optional-chaining) which would directly impact the functionality of this feature in TypeScript.

As of the last meeting of the ECMAScript standards committee, the proposal is staying at [Stage 1](https://tc39.github.io/process-document/) as it has some very fundamental questions about how it would be implemented.  While not a hard and fast rule, TypeScript will only implement Stage 3 proposals.  It sometimes implements Stage 2 proposals if they believe it is of critical importance and potentially usage in TypeScript drives the evolution of the standard.

I am not sure how any more clear people can be about that.

**Comment by mhegazy** [comment]
As I said [earlier](https://github.com/Microsoft/TypeScript/issues/16#issuecomment-290294158) this is on our short list. we are waiting on TC39 to reach some sort of consensus on the semantics of the operator. we would hate to put it out and then break users.

**Comment by notsonotso** [comment]
From TC39

> (null)?.b is undefined.

Oh my fucking god, gotta love Javascript. Not.

**Comment by RyanCavanaugh** [comment]
# This is not the thread to rehash the TC39 discussion

**Comment by mrpmorris** [comment]
I created something simple that meets my current needs. It will only work on a chain where each link is a property name, so accessing an element in an array (for example) isn't supported.

[Implementing a really simple Elvis Operator in TypeScript](https://mrpmorris.blogspot.co.uk/2017/09/implementing-really-simple-elvis.html)

**Comment by tolgaek** [comment]
Also if you have lodash/underscore, you can already use `_.get(Book, 'author.name.firstName')` which will do what this wants

**Comment by BjornMelgaard** [comment]
@tolgaek , `_.get` has bad typings, even with [this better typings](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/19433#issuecomment-326250652) ([not yet merged](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/19356), because of authors) typescript can definitely deduce type of result only if depth of object is 1, in all other cases it's `any` and must be checked in runtime

On other hand with elvis operator typescript may be able to infer type of result in objects with any depth, this is why Im looking forward to elvis operator

**Comment by tolgaek** [comment]
oh I see, I didn't know there was typings issue. Thanks @BjornMelgaard

**Comment by chaowlert** [comment]
@mhegazy can't this feature be implemented first and mark as experimental feature? I think people should not have problem if there is spec changed in experimental feature.

**Comment by notsonotso** [comment]
Elvis is not amused waiting this long.

**Comment by gs-akhan** [comment]
It has landed in Babel7. Typescript, we are running slow.. Someone please make this happen.
:)

**Comment by alangpierce** [comment]
@gs-akhan the Babel plugin implements an old version of the proposal from a few months ago. There have been changes to the proposal since then (including a significant change to how the operator is parsed), and there likely will be more changes before the feature reaches stage 2 (let alone stage 3), so any code written with the current babel plugin could break when the actual feature is released. Babel intentionally implements proposed features before they are stable so that the spec authors and other interested parties can try out the feature as proposed. Just because Babel has implemented a feature doesn't mean it can be implemented in a way that won't require breaking changes in the future.

**Comment by gs-akhan** [comment]
@alangpierce  That makes sense. Thanks

**Comment by RyanCavanaugh** [comment]
I get that this is a really, really nice operator, but having it be available before its rules have been ironed out is a footgun and we're not going to do that. The runtime behavior of the operator is still in flux; if you write code today it might stop working tomorrow in ways that aren't immediately apparent - maybe rare crashes, maybe data corruption, who knows? A little patience now will save you a *lot* of pain a few months down the road.

**Comment by atrauzzi** [comment]
Starting to think this ticket should be locked (not closed) until the spec is clear.

**Comment by oliverjanik** [comment]
When is the spec going to be ready?

**Comment by gisenberg** [comment]
@oliverjanik The current draft specification can be found [here](https://tc39.github.io/proposal-optional-chaining/). There's an agenda item to advance the proposal to stage 2 at the [next TC39 meeting](https://github.com/tc39/agendas/blob/master/2017/09.md) (9/26-9/28). I'll be presenting [these slides](https://docs.google.com/presentation/d/1iiCtJSW42Z7lg0YlagOZOfI-FTrkN88OuiO3pySawSo/edit?usp=sharing) at that time. For folks who'd like to review early and provide feedback, please file issues on the [proposal repo](https://github.com/tc39/proposal-optional-chaining).

**Comment by markwhitfeld** [comment]
Thank you so much @gisenberg for championing this issue for us! I was thinking about putting together a summary slide deck to help clear up the options around the operator that could be used at the TC39 meeting as to avoid confusion, but you have done this already. Awesome work!
Maybe just one other thing that could be beneficial to the TC39 conversation (and slide deck) is to look at the semantics and syntax of the operator in other languages. Although other languages should not necessarily dictate how it should work in Javascript, it would be helpful to keep the operator similar to that of other languages as to avoid confusion.
Good luck for next week!!!

**Comment by noppa** [comment]
Sorry for going a bit off topic once again, but I figured some folks here could find it interesting that in Flow it's now possible to add somewhat working type definitions for safe getter functions like `_.get`.

Example: [flowtype.org](https://flow.org/try/#0C4TwDgpgBACgTgezFAvFAzsOBLAdgcwB9cBXAWwCMI4BuAKDoBMIBjAGwEM5oAzE3FsGwJcUfBGAAeACoAuKAHkKAK1bAANFACC8gCQBpCCHQyAfKYAUCFfOmaA1kflaAlPID8ugKJsIZCLjA0uAQMppapvTM7Fy8-ILCouJScooqauHy8EiaAEJZiGCW1sq24XluUJ4+fgFBIZK6AHIiALIcIFTBkI01-oHdoXba5nmRTKyc3FB8AkIiYhIy8kqqgpmwhXkFOVAAwjtFVjZQw1p5mnuV1b799T3NbR1dDd63dYONLbjtnRCfb1qAwaZ1GUFyYL242iUzic0SixSK3S620h22m12B0xYE0ABFDsUTmcLvt8ddAXcAd9fi8Hn0Pq8ac9-q8GcCHsy-gD2fchuEwRDTJcwXjxnRQJAoHiOMAOKgoABvOgASA47nkyqgUBVKoo8lIlGo9G1upYHhlcoA2gBdegqgC+6jopsY8kwOAI9AdUUmsSgADcuFASvJLRx6HQWCJMFBcB5DVQ4ArksdlJoAOQcDOZigZlz0aO4WPoDwevD4FMSNOZ7OZlg5qAABkzjHz9CAA)

Not the prettiest piece of code ever and it doesn't correctly distinguish between null and undefined, but other than that it seems to work quite nicely.

AFAIK the only thing that's missing from TS for it to do the same is something like that `$NonMaybeType`.  
Not that it would remove the need for this operator, of course, I just thought that was cool.

**Comment by RyanCavanaugh** [comment]
This did not reach Stage 2 at the latest TC39 meeting due to concerns about syntactic consistency regarding bracket vs dot vs call access and semantic behavior (predictability around side effects of expressions to the right of an `undefined`/`null`-evaluated operand).

(vote 🎉 on this comment to throw rotten fruit)

**Comment by oliverjanik** [comment]
Thanks for letting us know. What a bummer. Maybe the scope needs to be narrowed so it's simpler but still useful?

**Comment by kitsonk** [comment]
> Maybe the scope needs to be narrowed so it's simpler but still useful?

This is the challenge that TC39 faces, which, while sucky, I have to admit I am glad people are going through this.  It is really hard for them to introduce a fairly complex level syntax and in fact, in this case the weak typing of the language is actually causing a significant amount of edge cases that have to be addressed, or you get code that goes 💥 which isn't good for anyone.  I don't think if you introduce an operator like this you can actually narrow its scope.  It would be easier for the implementors to cover 90% of the cases, but I don't think `¯\_(ツ)_/¯` is valid code for the remaining 10%.

**Comment by notsonotso** [comment]
Guys, 3+ years later, I'd say it's time to say "fuck you" to the committee and do it our own way, whatever is idiomatic to TypeScript. This will never work properly without proper typing. 

Make a TypeScript implementations with a syntax that is explicitely incompatible with the TC39 proposition. Once ES gets a safe-nav operator, TypeScript will have two options: one shitty-semantics that's compatible with ES and one that is tied to the type system. This means that you cannot use the TS safe-nav operator with *any*, and that's just fine.

**Comment by zlumer** [comment]
@notsnotso Typescript is not supposed to break JS, that's what Coffeescript is for.

**Comment by goloveychuk** [comment]
Maybe good solution is to:
- implement very simple ? operator for impetient devs, as experimental feature (like decorator), with warning - it WILL brake your code in future
- wait another 3 years when standart will be written, implement as not-experimental feature. Write tutorial, what could be broken.
"do it our own way, whatever is idiomatic to TypeScript" is not a case, beacuse in 3 years people will face with problem, that ts elvis doesn't work same way as in js.

**Comment by jehugaleahsa** [comment]
I am not surprised this did not go though, to be honest. In my original
post, I pointed out the ambiguity regarding indexing, function calls, etc.
Honestly, the more I think about `?.`, the more it feels like it doesn't
belong in JavaScript, since it has `undefined`.

I would much rather a more general-purpose utility that solved this problem
and more. One thought I had was something like an inline `try/catch`
expression (`let x = try a.b.c else 0`) in conjuction with an operator that
checked for "nully " (e.g., x ?? 1) rather than "falsey" (e.g., x || 1).
So, you'd combine them like this: `try a.b.c ?? 0 else 0`. It's wordy, yes,
but it basically says: Try to evaluate a.b.c and if the result is `null` or
`undefined`, return 0. If a or b is `undefined` and an exception is thrown,
catch it and return a 0.

An alternative would be to make the `else` clause optional, defaulting to
`undefined`. Then you could write the expression as: `let x= (try a.b.c) ??
0`. That's pretty darn compact, avoids the ambiguity and provides a more
general-purpose solution that can solve for other problems.

I'm not saying we should push for this instead; all I am saying is there
are alternatives to `?.` and we should really explore our options until we
find one that fits well with the JavaScript language.

**Comment by notsonotso** [comment]
@zlumer nice joke. That said, a TS specific feature would not break Javascript.

**Comment by notsonotso** [comment]
@jehugaleahsa I like your proposal, and it's found in other languages. I think it would work just fine for TS.

**Comment by MgSam** [comment]
I will say I don't really understand the strong need to wait until ECMAScript accepts the operator. TypeScript added classes, modules, lambdas, all long before they found their way into ECMAScript. In fact one of the stated goals of TypeScript was/is piloting experimental JS features. TypeScript providing its own implementation would no doubt help inform these discussions at the committee.

You guys have been willing to take breaking changes before when the committee eventually goes a different direction, I don't really see why this feature is so different.

**Comment by kgtkr** [comment]
async/await was also introduced to TypeScript at an early stage.

**Comment by trotyl** [comment]
> That said, a TS specific feature would not break Javascript.

There won't be something like "TS specific feature", please check [TypeScript Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals) for more information.

<sub>TypeScript arrives before ES going to a bright path, there could be some existing ones,
 they're just kept for compatibility.</sub>

**Comment by kitsonk** [comment]
> TypeScript added classes, modules, lambdas, all long before they found their way into ECMAScript. 

And modules in particular are a huge debacle that still cause no end to confusion.  That is an example that the TypeScript team have used repeatedly as an example of a _mistake_ and jumping the gun too early.  We also have private fields looming now, which again, I have been thankful for `private` for the past 5 years but it is going to cause no end to confusion.

Again, decorators have been available under the flag, but now that decorators are actually getting to Stage 3, it will require a re-implementation and code breakage for TypeScript.  That is why these things _cannot_ be taken likely.

> async/await was also introduced to TypeScript at an early stage.

Once the ECMAScript proposal reached Stage 3.

**Comment by RyanCavanaugh** [comment]
No one's misbehaving but if this conversation can't go anywhere but in circles (what if there were a flag? Yes we are aware of flags) or off-topic (what if TS stops being JS? No five years later we are not changing our mind on that), we don't need to be having it. We've said for about 3 years now that we will implement this exactly when the ES committee locks down its semantics.

Once again, the proposal repo is https://github.com/tc39/proposal-optional-chaining and you can follow its progress there. We'll be working offline to try to improve the proposal's chances at the next TC39 meeting because we really want this to go through too.

**Comment by DanielRosenwasser** [comment]
**Update:** We reached stage 2 this afternoon at TC39!!!

**Comment by RyanCavanaugh** [comment]
# Optional Chaining is Stage 3

Briefly unlocking this one *only* for celebratory purposes

**Comment by tikotzky** [comment]
![](https://media.giphy.com/media/26n6R5HOYPbekK0YE/giphy.gif)

**Comment by atrauzzi** [comment]
Hooray!

**Comment by CaLxCyMru** [comment]
![Safety First!](https://user-images.githubusercontent.com/3188575/61899259-008d2400-af13-11e9-9f34-9c6b4a5d0d70.gif)

**Comment by riskers** [comment]
Don't spam...

You can send an emoji expressing your feelings

**Comment by DanielRosenwasser** [comment]
Can somebody unsubscribe me from this thing?

**Comment by Zarel** [comment]
@DanielRosenwasser In case you're joking, or for anyone else who wants to unsubscribe and doesn't know how, you're looking for this button in the right sidebar:

![image](https://user-images.githubusercontent.com/551184/61900526-6f528900-aee3-11e9-86b4-ae8db33a465c.png)

**Comment by bgmort** [comment]
Not to mention there's a link in the email:

![image](https://user-images.githubusercontent.com/808854/61900942-dcfab700-aedb-11e9-89a7-dbed4500531d.png)

**Comment by DanielRosenwasser** [comment]
<sub>It was a joke, I'm working on the proposal.</sub>

**Comment by third774** [comment]
🎉🎉🎉

**Comment by charles-toller** [comment]
@RyanCavanaugh after unlocking this issue:
![martian](https://user-images.githubusercontent.com/7536966/61903699-03b7ee00-aeda-11e9-9d85-0c40a4446ec0.gif)

**Comment by coreh** [comment]
Really excited to see this finally land! 🎈 🎉 

Can't wait for the matching VSCode quick fix 😆 

<img width="553" alt="image" src="https://user-images.githubusercontent.com/418473/61907398-c0727500-af03-11e9-9263-d030fe94dacb.png">

**Comment by amritk** [comment]
![image](https://user-images.githubusercontent.com/2039539/61907681-cc017400-aee2-11e9-92a9-94f99e47e79e.png)

**Comment by kitsonk** [comment]
I am going to @ @RyanCavanaugh because he is probably unsubscribed to this thread and I want to be rude!  (and @DanielRosenwasser for good measure)

**Comment by jhpratt** [comment]
@kitsonk Don't be an ass.

**Comment by MatthiasKunnen** [comment]
@kitsonk, Why would RyanCavanaugh or DanielRosenwasser be unsubscribed to this thread? Ryan unlocked this issue and Daniel responded three comments above you.

Even if they were unsubscribed there is no need to cause more notification fatigue by spamming them.

**Comment by kbtzr** [comment]
Apparently GitHub is a terrible place for sarcastic humor, gee...

Big thanks to our champions at TC39 for figuring out the nasty details of this new language feature!

![thanks](https://user-images.githubusercontent.com/1047033/61916519-3c2deb00-af1f-11e9-8449-4138fae4a8bd.gif)

**Comment by DanielRosenwasser** [comment]
I think @kitsonk was just joking around, just as I sort of was. While we're being a little silly around the excitement, this is a gentle reminder to keep things civil as per the CoC.

**Comment by Kingwl** [comment]
@DanielRosenwasser 
Could I have to take a try on this? Or @rbuckton  has yet another existed branch for it 🙋🏻‍♂️

**Comment by rbuckton** [comment]
Yeah, that's fairly out of date, unfortunately.

**Comment by axmrnv** [comment]
I've just realized this thread started 5 years ago. :astonished:

**Comment by Kingwl** [comment]
@rbuckton 

> Yeah, that's fairly out of date, unfortunately.

Could I have a try on this?

**Comment by Losses** [comment]
![](https://tse1.mm.bing.net/th?id=OGC.da5f2ae5f52f3c1e65412149046d3e54&pid=Api&rurl=https%3a%2f%2fmedia.giphy.com%2fmedia%2fP8MxmGnjmytw s%2f200.gif&ehk=Drqx1%2fqgArrMmbpV%2bpIU8w)

**Comment by kitsonk** [comment]
Sorry @jhpratt @MatthiasKunnen I forgot we don't all share the same context on GitHub.  I've been bouncing around here for a long while and have spent time with Ryan and Daniel IRL and briefly participated in the recent event that gave rise to my misunderstood inside joke.  Apologies.

This whole issue though shows an interesting archeology of the design principles of TypeScript.  Ryan raised it at a time where TypeScript _might_ have actually considered syntax that pre-dated serious consideration in ECMAScript.  It was around that time though where TypeScript internally was learning some lessons of predicting ES2015 syntax that were having serious impacts on the language.  The team decided to not consider inclusion until there was a TC39 Stage 3 proposal.

While @RyanCavanaugh can address it, and I know he has been close to what has happened with the proposal, I suspect it would have been unlikely what the team would have implemented back in 2014 would have been 100% compatible with the current Stage 3 proposal.  So while it certainly is a celebration, also be thankful that we don't have 5 years of TypeScript code out that with a safe navigation operator that would not be entirely consistent with the behaviour in the proposal.

**Comment by TrevorSayre** [comment]
🙏

**Comment by kristinruben** [comment]
![westwing](https://user-images.githubusercontent.com/17237479/61954200-4cc57c00-af86-11e9-87e1-aee64f467083.gif)

**Comment by vitorhsb** [comment]
Is this the moment to remove [Waiting for TC39 label](https://github.com/microsoft/TypeScript/labels/Waiting%20for%20TC39)? 🤔

**Comment by smeijer** [comment]
> Is this the moment to remove [Waiting for TC39 label](https://github.com/microsoft/TypeScript/labels/Waiting%20for%20TC39)? 🤔

And add `ship-it`

**Comment by aexol** [comment]
wowowow

**Comment by soulman-is-good** [comment]
Finally. Literally was asking around y'day when will it be stage 3, and bam! Thanks everyone who were working to push it through!

**Comment by opnksyn** [comment]
How to mute this thread? :)

**Comment by styfle** [comment]
@opnksyn If you don't care about the spammy excitement, there is a feature to get notified when this issue is closed. This will prevent all the comments from showing up in notifications like the one you just made 😄 

![image](https://user-images.githubusercontent.com/229881/61968556-55c64580-afa6-11e9-8e23-4bca200a0d83.png)

**Comment by be5invis** [comment]
![image](https://user-images.githubusercontent.com/240091/61985582-a6e23380-afbe-11e9-9b2f-176666621ec3.png)

**Comment by HitalloExiled** [comment]
Any emit solution already defined?  
Something like this could be interesting:

```ts
function __chain<T extends object, U>(value: T|null|undefined, callback: (value: T) => U): U|undefined {
    if (value !== null && value !== undefined) {
        return callback(value);
    }

    return undefined;
}

type Foo = { x?: { y?: { z?: () => number } } }

const foo: Foo = { }

// x.?y.?z?()
const value = __chain(foo, _a => __chain(_a.x, _b => __chain(_b.y, _c => __chain(_c.z, _d => _d()))));
```

**Comment by ExE-Boss** [comment]
I prefer what Babel does as it avoids creating unnecessary function scopes:
```js
var _foo, _foo$x, _foo$x$y, _foo$x$y$z;

// foo?.x?.y?.z?.()
(_foo = foo) === null || _foo === void 0 ? void 0
	: (_foo$x = _foo.x) === null || _foo$x === void 0 ? void 0
	: (_foo$x$y = _foo$x.y) === null || _foo$x$y === void 0 ? void 0
	: (_foo$x$y$z = _foo$x$y.z) === null || _foo$x$y$z === void 0 ? void 0
	: _foo$x$y$z.call(_foo$x$y);
```

**Comment by jonathanmarvens** [comment]
> I prefer what Babel does as it avoids creating unnecessary function scopes:
> 
> ```js
> var _foo, _foo$x, _foo$x$y, _foo$x$y$z;
> 
> // foo?.x?.y?.z?.()
> (_foo = foo) === null || _foo === void 0 ? void 0
> 	: (_foo$x = _foo.x) === null || _foo$x === void 0 ? void 0
> 	: (_foo$x$y = _foo$x.y) === null || _foo$x$y === void 0 ? void 0
> 	: (_foo$x$y$z = _foo$x$y.z) === null || _foo$x$y$z === void 0 ? void 0
> 	: _foo$x$y$z.call(_foo$x$y);
> ```

I agree, @ExE-Boss. I feel that avoiding creating unnecessary function scopes is ideal (even if it makes the compiled code a bit ugly)

**Comment by gwardwell** [comment]
Performance and adherence to the ES specification should definitely win out over readability when we’re talking about compiled code.

**Comment by proteria** [comment]
Is there any reason the code compiled by Babel compares to both `null` and `void 0` with triple equals instead of a simple `== null`?

**Comment by jonathanmarvens** [comment]
> Is there any reason the code compiled by Babel compares to both `null` and `void 0` with triple equals instead of a simple `== null`?

@proteria I just took a quick look at [the optional chaining code for __Babel__](https://github.com/babel/babel/blob/4f0840ab88037fb0241a745f7e4b0f8969709ef3/packages/babel-plugin-proposal-optional-chaining/src/index.js); it seems that if you pass in the `loose` option and set it to a truthy value, it won’t produce the strict equality checks:

https://github.com/babel/babel/blob/4f0840ab88037fb0241a745f7e4b0f8969709ef3/packages/babel-plugin-proposal-optional-chaining/src/index.js#L8

https://github.com/babel/babel/blob/4f0840ab88037fb0241a745f7e4b0f8969709ef3/packages/babel-plugin-proposal-optional-chaining/src/index.js#L99-L113

**Comment by noppa** [comment]
That's because of [`document.all`](https://html.spec.whatwg.org/multipage/obsolete.html#dom-document-all), a quirk in the language that gets special treatment in the language for backwards compatibility.

```javascript
document.all == null // true
document.all === null || document.all === undefined // false
```
In the optional chaining proposal
```javascript
document.all?.foo === document.all.foo
```
but `document.all == null ? void 0 : document.all.foo` would incorrectly return `void 0`. In loose mode, this detail of the spec is dismissed for simplicity/performance/generated code size as most people don't have to deal with `document.all` anyway.

**Comment by jhpratt** [comment]
Surely the `document.all` case could be special? It shouldn't require much extra code, just a few lines to check the object and property.

**Comment by ExE-Boss** [comment]
Except that `document.all` can&nbsp;be&nbsp;assigned to&nbsp;a&nbsp;variable, and&nbsp;tracking where it’s&nbsp;used requires a&nbsp;type system.

**Comment by jhpratt** [comment]
I'm aware of that. Babel doesn't have a type system, TypeScript does. Perhaps it's not as simple as I made it sound, but I imagine there's already code for certain situations that is capable of tracking usage.

**Comment by G-Rath** [comment]
Actually, you don't need a type system to track that - the special behaviour is actually on [`HTMLAllCollection`](https://html.spec.whatwg.org/multipage/common-dom-interfaces.html), not [`document.all`](https://html.spec.whatwg.org/multipage/obsolete.html#dom-document-all).

So you should just be able to do an `instanceof HTMLAllCollection` check, and you'll be golden.

**Comment by noppa** [comment]
Yeah but... why'd you do `instanceof` when you can just do `=== null || === void 0`? Surely that's more simple.

**Comment by G-Rath** [comment]
For sure - I was just pointing out you don't need a type system to track `document.all` :)

Personally, imo just break it and fix it if someone complains - `document.all` is meant to be deprecated, and given that optional chaining is pretty much the extreme opposite of deprecated ("new and shiny"), the chances of using the two together are very slim.

TypeScript is pretty popular, so it would also serve as another nail in `document.all`s coffin, to encourage people to stop using it.

Plus it's pretty easy to work around: "manually" chain on that property - you can still chain everything else just fine.

That's my 2-cents on the matter, but it doesn't bother me if it's supported or not.

tl;dr: I was just pointing out you didn't need a type system to identify `document.all` :joy:

**Comment by jhpratt** [comment]
@noppa It can be performed at compile time. If `foo instanceof HTMLAllCollection` is `true`, emit `foo === null || foo === void 0`, otherwiise we can _safely_ emit `foo == null`.

@G-Rath Like it or not, deprecated doesn't mean it shouldn't work. TypeScript should remain compatible with JavaScript.

**Comment by ExE-Boss** [comment]
@jhpratt But that currently goes against the&nbsp;[TypeScript Design Non&#x2011;Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals#non-goals).

Also, you’d still have to&nbsp;do&nbsp;`foo === null || foo === void 0` for&nbsp;anything to&nbsp;which `HTMLAllCollection` could be&nbsp;assigned, eg.&nbsp;`any`&nbsp;or&nbsp;`object`, so&nbsp;I don’t think it’s&nbsp;really worth&nbsp;it.

**Comment by jhpratt** [comment]
I presume you're referring to non-goal (5)

> Add or rely on run-time type information in programs, or emit different code based on the results of the type system. Instead, encourage programming patterns that do not require run-time metadata.

Though I agree this would certainly emit different code based on the type, it's only to reduce code size. Though as you've pointed out, it's not quite as simple as checking for `HTMLAllCollection`.

To be fair, TS _has_ rejected a potential minifier that uses type information, and this is (sort of) related — the primary reason to emit `== null` is to reduce code size based on type information.

If this _isn't_ implemented, it would be great if the lang team adds an option to tsconfig similar to Babel's "loose" option.

**Comment by fbartho** [comment]
>If this isn't implemented, it would be great if the lang team adds an option to tsconfig similar to Babel's "loose" option.

Relatedly, many of us use TypeScript for build tools, and for mobile applications, none of which have to worry about browser constraints!

In fact, we use both TS & Babel together, so maybe one of these options should be to passthrough the operator to Babel/underlying runtime!

**Comment by kibertoad** [comment]
Optional chaining is Stage 3 now, so implementing it seems like a safe bet at this point: https://github.com/tc39/proposal-optional-chaining

**Comment by Zarel** [comment]
@fbartho 

> In fact, we use both TS & Babel together, so maybe one of these options should be to passthrough the operator to Babel/underlying runtime!

I don't understand this comment. You don't need any extra options to passthrough the operator to Babel; if you have TypeScript set up for Babel, you already have `noEmit: true` which already passes _everything_ through to Babel.

**Comment by fbartho** [comment]
@Zarel Babel’s TypeScript implementation is missing several features that our codebase was already relying on, including namespaces and const enums. We’re using TSC with emit enabled, and applying Babel as a second transformation. (We’re working on getting rid of the namespaces, but it’s unclear if we’ll ever be able to get rid of all of the mismatched features)

**Comment by WesSouza** [comment]
As a note, the [Optional Chaining for JavaScript](https://github.com/tc39/proposal-optional-chaining) proposal reached stage 3 recently.

**Comment by RyanCavanaugh** [comment]
People coming to this thread should start at the [earlier stage 3 announcement](https://github.com/microsoft/TypeScript/issues/16#issuecomment-515160784) and read the comments starting there (blame GitHub for hiding tons of user content with no straightforward way to load everything)

**Comment by gastonmorixe** [comment]
❤️

**Comment by domske** [comment]
Great feature - "optional chaining" / "safe navigation". Especially in TypeScript strict-mode. Awesome to hear that this will be implemented soon. ❤️ 

This brought me here and I hope it will be supported. Just an use case:

Expected in TypeScript 3.7.
```ts
document.querySelector('html')?.setAttribute('lang', 'en');
```

*VS* 

Currently in TypeScript 3.5.
```ts
const htmlElement = document.querySelector('html');
if (htmlElement) {
  htmlElement.setAttribute('lang', 'en');
}
```

Will this work without any errors? Or is this still a `TypeError: Cannot read property 'setAttribute' of null.`? The `?` op. should be cancel further chains after null / undefined.

```ts
class Test {
  it() {
    console.log('One');
    document.querySelector('html')?.setAttribute('lang', 'en');
    console.log('Two');
  }
}
```
I expect following: 
If html element does not exist (null). The console should logged `One` and `Two`, and the `setAttribute` method is not tried to invoked. (No errors). 
Did I understand that correctly?

**Comment by joehillen** [comment]
FYI, this isn't strictly a TS feature; it's a JS feature.

According to the [TC39 proposal](https://github.com/tc39/proposal-optional-chaining) the syntax will be:

```ts
document.querySelector('html')?.setAttribute?.('lang', 'en');
```

**Comment by RyanCavanaugh** [comment]
The discussion has started to go circular again so we're back to locked state.

I truly beg anyone tempted to leave a comment in a 100+-comment long GitHub thread to really commit to reading all the prior comments first. Probably your question, and the answer to it, will be found there!
